src/graph_compute/numa_scheduling.h

```diff
@@ -138,107 +178,184 @@ static void createChunkData(vertex_t * const nodes, const vid_t cntNodes,
+                      vid_t queueNumber =
+                        scheddata->chunkdata[enabledChunk].workQueueNumber;
+                      numaSchedInit[queueNumber].workQueue->push(enabledChunk);
+                    }
+                  }
                 }
+                chunkdata->nextIndex++;
+              }
+            }
+            if (chunkdata->nextIndex == chunkdata->phaseEndIndex[phase]) {
+              //  this chunk is at the end of the phase
+              doneFlag = true;
+              if ((phase + 1) == config->numPhases) {
+                //  we finished the entire chunk (i.e., all phases)
+                //  so, we reset the index for next round
+                chunkdata->nextIndex = (chunk << CHUNK_BITS);
+              }
+              if (__sync_sub_and_fetch(config->remainingChunks, 1) == 0) {
+                //  this was the last chunk for this phase
+                //  so, reset the remainingChunks counter and
+                //  advance phase counter (mod numPhases)
+                *config->remainingChunks = config->scheddata->cntChunks;
+                (*config->phase)++;
+                if (*config->phase == config->numPhases) {
+                  *config->phase = 0;
+                }
+                //  make sure writes to these shared variables are visible
+                __sync_synchronize();
               }
-            } else {
-              scheddata->chunkdata[i].nextIndex = j;
-              localDoneFlag = true;  // we couldn't process one of the nodes, so break
-              doneFlag = false;  // we couldn't process one, so we need another round
             }
-            j++;
-          }
-          if (!localDoneFlag) {
-            scheddata->chunkdata[i].nextIndex = j;
           }
         }
       }
     }
   }
+  return NULL;
+}
+
+static inline void execute_rounds(const int numRounds,
+                                  vertex_t * const nodes,
+                                  const vid_t cntNodes,
+                                  scheddata_t * const scheddata,
+                                  global_t * const globaldata) {
+  //  the shared variable for tracking remaining chunks during a phase
+  volatile int remainingChunks = scheddata->cntChunks;
+  //  the shared variable for tracking the phase of the round
+  volatile int phase = 0;
+  #pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+  for (int i = 0; i < NUMA_WORKERS; i++) {
+    scheddata->numaSchedInit[i].globaldata = globaldata;
+    scheddata->numaSchedInit[i].numRounds = numRounds;
+    scheddata->numaSchedInit[i].remainingChunks = &remainingChunks;
+    scheddata->numaSchedInit[i].phase = &phase;
+  }
+  pthread_t * workers =
+    static_cast<pthread_t *>(malloc(sizeof(pthread_t)*NUMA_WORKERS));
+  for (int i = 0; i < NUMA_WORKERS; i++) {
+    numaSchedInit_t * config = &scheddata->numaSchedInit[i];
+    assert(pthread_create(&workers[i], NULL, processChunks,
```

Comment:
`assert(x)` type statements get completely compiled out if assertions are disabled with `NDEBUG`, side-effects and all. Try enabling `NDEBUG` and you'll see that no threads will be created.

Please move the thread creation outside the assertion, and simply assert on the result.